prime-router/src/main/kotlin/Submission.kt

```diff
@@ -76,6 +78,114 @@ class DetailedSubmissionHistory(
      */
     var externalName: String? = null
 
+    /**
+     * The step in the delivery process for a submission
+     * Supported values:
+     *     ERROR - error on initial submission
+     *     RECEIVED - passed the received step in the pipeline and awaits processing/routing
+     *     NOT_DELIVERING - processed but has no intended receivers
+     *     WAITING_TO_DELIVER - processed but yet to be sent to/downloaded by any receivers
+     *     PARTIALLY_DELIVERED - processed, successfully sent to/downloaded by at least one receiver
+     *     DELIVERED - processed, successfully sent to/downloaded by all receivers
+     * @todo For now, no "send error" type of state.
+     *     If a send error occurs for example,
+     *     it'll just sit in the waitingToDeliver or
+     *     partiallyDelivered state until someone fixes it.
+     */
+    enum class Status(val printableName: String) {
+        ERROR("Error"),
+        RECEIVED("Received"),
+        NOT_DELIVERING("Not Delivering"),
+        WAITING_TO_DELIVER("Waiting to Deliver"),
+        PARTIALLY_DELIVERED("Partially Delivered"),
+        DELIVERED("Delivered"),
+    }
+
+    /**
+     * Summary of how far along the submission's process is.
+     * The supported values arre listed in the Status enum.
+     */
+    val overallStatus: Status get() {
```

Comment:
You use this overallStatus all over the place, so having it as a getter is probably inefficient.  

You could do it as a lateinit. 

But there is an implicit ordering to the calculations that need to happen:   the two date fields depend on the overallstatus which depends on the realdestinations.    

So it might be wise to make that ordering more explicit.   A clean way to do this is with an init { } clause that gets called during construction.   An example of that is found in AuthenticatedClaims.kt, where an init { } is used to initialize a bunch of class fields during construction.